haystack_experimental/core/pipeline/async_pipeline.py

```diff
@@ -2,17 +2,592 @@
+        scheduled_components: Set[str] = set()
+
+        # 2) Convert input data
+        prepared_data = self._prepare_component_input_data(data)
+        self._validate_input(prepared_data)
+        inputs_state = self._convert_from_legacy_format(prepared_data)
+
+        # For quick lookup of downstream receivers
+        ordered_names = sorted(self.graph.nodes.keys())
+        cached_receivers = {
+            n: self._find_receivers_from(n) for n in ordered_names
+        }
+
+        # We fill the queue once and raise if all components are BLOCKED
+        self.validate_pipeline(self._fill_queue(ordered_names, inputs_state))
+
+        # Single parent span for entire pipeline execution
+        with tracing.tracer.trace(
+            "haystack.async_pipeline.run",
+            tags={
+                "haystack.pipeline.input_data": data,
+                "haystack.pipeline.metadata": self.metadata,
+                "haystack.pipeline.max_runs_per_component": self._max_runs_per_component,
+            },
+        ) as parent_span:
+
+            # -------------------------------------------------
+            # We define some functions here so that they have access to local runtime state
+            # (inputs, tasks, scheduled components) via closures.
+            # -------------------------------------------------
+            async def _run_component_async(
+                component_name: str,
+                component_inputs: Dict[str, Any],
+            ) -> Dict[str, Any]:
+                """
+                Runs one component.
+
+                If the component supports async, await directly it will run async; otherwise offload to executor.
+                Updates visits count, writes outputs to `inputs_state`,
+                and returns pruned outputs that get stored in `pipeline_outputs`.
+
+                :param component_name: The name of the component.
+                :param component_inputs: Inputs for the component.
+                :returns: Outputs from the component that can be yielded from run_async_generator.
+                """
+                comp_dict = self._get_component_with_graph_metadata(component_name)
+                if comp_dict["visits"] >= self._max_runs_per_component:
+                    raise PipelineMaxComponentRuns(
+                        f"Max runs for '{component_name}' reached."
+                    )
+
+                instance: Component = comp_dict["instance"]
+                with tracing.tracer.trace(
+                    "haystack.component.run_async", parent_span=parent_span
+                ):
+                    logger.info("Running component {name}", name=component_name)
+
+                    if getattr(instance, "__haystack_supports_async__", False):
+                        outputs = await instance.run_async(**component_inputs)  # type: ignore
+                    else:
+                        loop = asyncio.get_event_loop()
```

Comment:
Can you clarify if here the intention is to create another another event loop if it doesn't exist. If not, we can also look into using `asyncio.get_running_loop()` which is recommended to prevent unexpected behaviour. 
Also asyncio.get_event_loop() is deprecated since python 3.12.
[Source](https://docs.python.org/3/library/asyncio-eventloop.html)
```suggestion
                        loop = asyncio.get_running_loop()
```

Reply:
Thanks, this looks good. This part of the code was the same as in Madeesh's original implementation.